tests/components/google_pubsub/test_init.py

```diff
@@ -0,0 +1,262 @@
+    )
+
+
+async def test_full_config(hass, mock_client):
+    """Test the full config of the component."""
+    config = {
+        google_pubsub.DOMAIN: {
+            "project_id": "proj",
+            "topic_name": "topic",
+            "credentials_json": "creds",
+            "filter": {
+                "include_domains": ["light"],
+                "include_entity_globs": ["sensor.included_*"],
+                "include_entities": ["binary_sensor.included"],
+                "exclude_domains": ["light"],
+                "exclude_entity_globs": ["sensor.excluded_*"],
+                "exclude_entities": ["binary_sensor.excluded"],
+            },
+        }
+    }
+    assert await async_setup_component(hass, google_pubsub.DOMAIN, config)
+    await hass.async_block_till_done()
+    assert hass.bus.listen.called
+    assert EVENT_STATE_CHANGED == hass.bus.listen.call_args_list[0][0][0]
+    assert mock_client.PublisherClient.from_service_account_json.call_count == 1
+    assert (
+        mock_client.PublisherClient.from_service_account_json.call_args[0][0] == "path"
+    )
+
+
+FilterTest = namedtuple("FilterTest", "id should_pass")
+
+
+def make_event(entity_id):
+    """Make a mock event for test."""
+    domain = split_entity_id(entity_id)[0]
+    state = mock.MagicMock(
+        state="not blank",
+        domain=domain,
+        entity_id=entity_id,
+        object_id="entity",
+        attributes={},
+    )
+    return mock.MagicMock(data={"new_state": state}, time_fired=12345)
+
+
+async def _setup(hass, filter_config):
+    """Shared set up for filtering tests."""
+    config = {
+        google_pubsub.DOMAIN: {
+            "project_id": "proj",
+            "topic_name": "topic",
+            "credentials_json": "creds",
+            "filter": filter_config,
+        }
+    }
+    assert await async_setup_component(hass, google_pubsub.DOMAIN, config)
+    await hass.async_block_till_done()
+    return hass.bus.listen.call_args_list[0][0][1]
+
+
+async def test_allowlist(hass, mock_client):
+    """Test an allowlist only config."""
+    handler_method = await _setup(
+        hass,
+        {
+            "include_domains": ["light"],
+            "include_entity_globs": ["sensor.included_*"],
+            "include_entities": ["binary_sensor.included"],
+        },
+    )
+    publish_client = mock_client.PublisherClient.from_service_account_json("path")
+
+    tests = [
+        FilterTest("climate.excluded", False),
+        FilterTest("light.included", True),
+        FilterTest("sensor.excluded_test", False),
+        FilterTest("sensor.included_test", True),
+        FilterTest("binary_sensor.included", True),
+        FilterTest("binary_sensor.excluded", False),
+    ]
+
+    for test in tests:
+        event = make_event(test.id)
+        handler_method(event)
+
+        was_called = publish_client.publish.call_count == 1
+        assert test.should_pass == was_called
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5745133)